Allow QuickUMLS to be used as component in spacy pipeline

burgersmoke opened the pull request:
This can be used as an entity matcher for UMLS concepts in other modular spacy pipelines.  This is already being used in an operational capacity for syndromic surveillance in a healthcare system.  More info on this can be provided upon request.

This is in response to this issue:
https://github.com/Georgetown-IR-Lab/QuickUMLS/issues/40

soldni commented on quickumls/spacy_component.py:
```diff
@@ -0,0 +1,90 @@
+            if self.verbose:
+                print('Total matches after best match: [{0}]'.format(len(matches)))
+            
+        self.quickumls._print_verbose_status(doc, matches)
```
I would prefer repeating the inner logic of the `QuickUMLS.match` here, as it means that, in case the core logic is changed, we'd have to track any modification in two places. Could you wrap all the shared logic between `QuickUMLS` and `SpacyQuickUMLS` into a single method (something like `QuickUMLS._match`) that gets called by `QuickUMLS.match` and `SpacyQuickUMLS.__call__`?

burgersmoke replied:
Great idea.  I had considered that change but I did not make it since I wasn't sure how comfortable you would be with that.  But now that I have your approval, I will make this change and update the PR.  This is exactly the feedback I wanted before I clean this up for the "true" PR.

burgersmoke replied:
OK, I updated this and this is working as expected both as QuickUMLS "standalone" and spacy component.  Can you take a look over this and ensure this matches your thinking?

soldni commented on spacy_example_pipeline.py:
```diff
@@ -0,0 +1,29 @@
+nlp.remove_pipe('tagger')
+nlp.remove_pipe('parser')
+
+# TODO -- change this from a hard coded path to a small chunk of UMLS that can be legally distributed
```
Unlikely we can distribuite any of UMLS, but maybe some files that looks enough like UMLS?

burgersmoke replied:
Completely agree.  No reason to keep a hard-coded path.  I considered this to be a "rough draft" PR and this item I had not considered.  I will update this.

burgersmoke replied:
Also, I did comment on the UMLS issue.  A colleague told me once that they **believed** that there was a  small subset of the UMLS that could be distributed.  I'll follow up on that.   In the meantime, I agree that this is not necessary to bring this PR into the repo.  That can be later work.  

Speaking of later work, I'll follow up with you and anyone else interested in what we are planning with a project called [medspacy](https://github.com/medspacy/medspacy).  We'd like to include QuickUMLS as part of the default pipeline that people can install with a single command.

burgersmoke replied:
OK, I re-read your comment and understood it.  I removed the previous example .py file and put a small example in the README.  I'm open to any adjustments for the README including order or placement.

burgersmoke commented:
Great.  I appreciate all of this feedback since I know this was a rough PR.  

One question: You mention moving spacy_example_pipeline.py.  Are you proposing that it goes into /quickumls or into another location?  Maybe something like /examples?

Now that I have your feedback, I'll make these changes tonight or Tuesday and update the PR.

ljak commented:
Went through it quickly, seems nice for first iteration ! Thanks for work.

burgersmoke commented:
OK, I think this is ready to go now with `b06b882`.  Besides the items above, I also added documentation, fixed some comments and fixed the hard-coded `QuickUMLS` arguments to be passed in as `**kwargs` instead.  Now I feel comfortable with this PR. :)

MichalMalyska commented on README.md:
```diff
@@ -52,6 +52,29 @@ Set `best_match` to `False` if you want to return overlapping candidates, `ignor
+
+for ent in doc.ents:
+    print('Entity text : {}'.format(ent.text))
+    print('Label (UMLS CUI) : {}'.format(ent.label_))
```
Why is this not ent._.label_ ?

burgersmoke replied:
Great question.  `label_` is already an [attribute of Span](https://spacy.io/api/span#attributes) which is already spaCy convention.  I didn't see a need to add another extension when there is already a member which exists for this need.  

In our systems that use these entities in downstream tasks, we are able to use `ent.label_` from a number of "upstream" NER components, whether they are QuickUMLS or scispacy or anything else.  In short, this allows better compatibility by minimizing extensions beyond spacy.

burgersmoke commented:
Any update on this?  I believe I've resolved the code issue and remaining questions.  Are there any other questions about this change I can answer?  I'm wanting to move this forward since we're planning to leverage this to make medspacy and QuickUMLS better with some mutually beneficial upcoming work.

soldni reviewed (approved):
Sorry for the delay in reviewing this! The new changes look good to me.

burgersmoke commented:
Fantastic.  Thank you @soldni !

soldni commented:
@burgersmoke thank you for your hard work! will push out a proper release (v.1.5) momentarily.

burgersmoke commented:
My pleasure.  There's more to come as QuickUMLS is an important part of our plans for medspacy.  I'll continue to follow up with you and anyone else interested.

soldni commented:
I was doing some tests and notice that, in the latest version of spaCy (2.3), I get an error if two entities overlap:

```python
Traceback (most recent call last):
  File "test.py", line 10, in <module>
    doc = nlp('Pt c/o shortness of breath, chest pain, nausea, vomiting, diarrrhea')
  File "/home/ubuntu/anaconda3/envs/quickumls/lib/python3.7/site-packages/spacy/language.py", line 449, in __call__
    doc = proc(doc, **component_cfg.get(name, {}))
  File "/home/ubuntu/qumls_1.4/QuickUMLS/quickumls/spacy_component.py", line 78, in __call__
    doc.ents = list(doc.ents) + [span]
  File "doc.pyx", line 553, in spacy.tokens.doc.Doc.ents.__set__
ValueError: [E103] Trying to set conflicting doc.ents: '(8, 10, 'C0008031')' and '(8, 10, 'C2926613')'. A token can only be part of one entity, so make sure the entities you're setting don't overlap.
```

According to this [GitHub issue](https://github.com/explosion/spaCy/issues/3608), this seems to be the expected behavior since at least spaCy (2.1.3)

I am planning to solve this by either (a) adding a custom extension type called `quickumls` to docs, or (b) have each span be a list of matches. Any preference? In the first case, you'd access matches as follows:

```python
for ent in doc._.quickumls:
    print(ent.text, ent.label_, ent._.semtypes, ent._.similarity)
```

In the latter, you'd use the following syntax:

```python
for ent in doc.ents:
    for match in ent._.quickumls:
        print(ent.text, match.cui, match.semtypes, match.similarity)
```

I personally prefer the second one, as it makes more sense to me to just have an entity with multiple labels, but please do let me know which one you think would make more sense.

burgersmoke commented:
You are absolutely correct and I debated raising it earlier, but I wanted to take steps towards the component being included first so that we could have this conversation.  I've opened this as a new issue here.  I'll add my initial thoughts here:

https://github.com/Georgetown-IR-Lab/QuickUMLS/issues/60
